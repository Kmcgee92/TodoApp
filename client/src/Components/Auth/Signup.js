import React, { useEffect, useState } from "react";
// redux
import { useDispatch } from "react-redux";
import { signupHandler } from "../../redux/actions/authActions";
// apollo
import { useMutation } from "@apollo/react-hooks";
// import { GET_USER } from "../../graphql/queries/GetUser";
import { GET_USER_BY_SIGNUP } from "../../graphql/mutations/signup";
// animxyz animation
import "@animxyz/core";
// mui components
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import LinearProgress from "@material-ui/core/LinearProgress";
// mui icons
import PersonOutlineIcon from "@material-ui/icons/PersonOutline";
import HttpsOutlinedIcon from "@material-ui/icons/HttpsOutlined";
import VisibilityOffOutlinedIcon from "@material-ui/icons/VisibilityOffOutlined";
import VisibilityOutlinedIcon from "@material-ui/icons/VisibilityOutlined";
// concat classes
import clsx from "clsx";



const Signup = ({ classes, setModalOpen }) => {
  const dispatch = useDispatch();
  const [passVisibility, setPassVisibility] = useState(false);
  const [confirmPassVisibility, setConfirmPassVisibility] = useState(false);
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const [nameError, setNameError] = useState(false);
  const [passwordError, setPasswordError] = useState(false);
  const [confirmPasswordError, setConfirmPasswordError] = useState(false);
  const [creatingUserLoader, setCreatingUserLoader] = useState(false);
  const [overThirty, setOverThirty] = useState(false);
  const [serverError, setServerError] = useState("");

  let [getUserBySignup, { data, loading, error }] = useMutation(
    GET_USER_BY_SIGNUP
  );

  useEffect(() => {
    setServerError("");
    if (!error && data && !loading) {
      if (data.Signup.error) {
        setServerError(data.Signup.error);
        return;
      }
      dispatch(signupHandler(data));
      setModalOpen(false);
    }
  }, [setModalOpen, dispatch, error, data, loading]);
  // Loading bar
  useEffect(() => {
    if (loading) {
      setCreatingUserLoader(true);
    }
    let timeoutName = setTimeout(() => setCreatingUserLoader(false), 2800);
    return () => clearTimeout(timeoutName);
  }, [loading, creatingUserLoader, setCreatingUserLoader]);

  // over 30 error
  useEffect(() => {
    setOverThirty(false);
    if (name.length > 30) {
      setOverThirty(true);
    }
  }, [name, overThirty]);

  // password and confirm requirements
  useEffect(() => {
    if (password.length >= 8) {
      setPasswordError(false);
    }
    setConfirmPasswordError(false);
    if (password !== confirmPassword && confirmPassword) {
      setConfirmPasswordError(true);
    }
  }, [password, confirmPassword, setPasswordError]);

  const handleSignup = async (e) => {
    e.preventDefault();
    setNameError(false);
    setPasswordError(false);
    let nameCurrErr = false;
    let passCurrErr = false;
    let confirmPassCurrErr = false;

    if (password.length < 8) {
      passCurrErr = true;
      setPasswordError(true);
    }
    if (confirmPasswordError) {
      confirmPassCurrErr = true;
    }

    const regex = /^[a-z ]{3,30}$/i;
    if (!regex.test(name)) {
      nameCurrErr = true;
      setNameError(true);
    }

    if (!nameCurrErr && !passCurrErr && !confirmPassCurrErr) {
      setCreatingUserLoader(true);
      await getUserBySignup({
        variables: {
          name: name,
          email: email,
          password: password,
        },
      });
    }

    setCreatingUserLoader(false);
  };

  const nameHelperText = () => {
    const handleOverThirty = clsx({
      [classes.nameLength]: !overThirty,
      [classes.nameLengthError]: overThirty,
    });
    if (nameError) {
      return (
        <>
          <span className={handleOverThirty}>{name.length} of 30</span>
          <span>
            must be between 3 and 30 characters and must not contain special
            characters `!@#$%&*` or numbers.
          </span>
        </>
      );
    } else {
      return <span className={handleOverThirty}>{name.length} of 30</span>;
    }
  };

  return (
    <>
      <div className={classes.modalContainer}>
        <div className="xyz-in" xyz="inherit down">
          <div className={classes.modalContent}>
            <header className={classes.signupHeader}>
              <h2>Create an Account</h2>
            </header>
            <span style={{ padding: "0 20px", color: "black" }}>
              * indicates required fields
            </span>
            {serverError ? (
              <div className={classes.serverErrorStyles}>
                <span className={classes.serverErrorStyles}>
                  {serverError}
                </span>
              </div>
            ) : null}
            <form onSubmit={handleSignup} className={classes.signupForm}>
              <TextField
                required
                data-test="nameField"
                onChange={(e) => setName(e.target.value)}
                error={nameError}
                helperText={nameHelperText()}
                label="Name"
                type="name"
                autoComplete="name"
                variant="outlined"
                className={classes.signupInput}
                InputProps={{
                  className: classes.signupInputChildren,
                  endAdornment: <PersonOutlineIcon />,     
                }}
                autoFocus={true}
                />
              <TextField
                required
                data-test="emailField"
                onChange={(e) => setEmail(e.target.value)}
                label="Email"
                type="email"
                autoComplete="email"
                variant="outlined"
                className={classes.signupInput}
                InputProps={{
                  className: classes.signupInputChildren,
                  endAdornment: <HttpsOutlinedIcon />,
                }}
                />
              <TextField
                required
                data-test="passwordField"
                onChange={(e) => setPassword(e.target.value)}
                error={passwordError}
                helperText={
                  passwordError &&
                  "password must be at least 8 characters long "
                }
                color="primary"
                label="Password"
                name="password"
                type={!passVisibility ? "password" : "reveal"}
                variant="outlined"
                autoComplete="currentPassword"
                className={classes.signupInput}
                InputProps={{
                  className: classes.signupInputChildren,
                  endAdornment: passVisibility ? (
                    <VisibilityOutlinedIcon
                    onClick={() => setPassVisibility(false)}
                    className={classes.visibilityIcon}
                    style={{ color: "white" }}
                    />
                    ) : (
                      <VisibilityOffOutlinedIcon
                      onClick={() => setPassVisibility(true)}
                      className={classes.visibilityIcon}
                      />
                      ),
                    }}
                    />
              <TextField
                required
                data-test="confirmPasswordField"
                onChange={(e) => setConfirmPassword(e.target.value)}
                error={confirmPasswordError}
                helperText={
                  confirmPasswordError &&
                  "confirmation does not match password field."
                }
                label="Confirm Password"
                type={!confirmPassVisibility ? "password" : "reveal"}
                variant="outlined"
                autoComplete="current-password"
                className={classes.signupInput}
                InputProps={{
                  className: classes.signupInputChildren,
                  endAdornment: confirmPassVisibility ? (
                    <VisibilityOutlinedIcon
                      onClick={() => setConfirmPassVisibility(false)}
                      style={{ color: "white" }}
                    />
                  ) : (
                    <VisibilityOffOutlinedIcon
                      onClick={() => setConfirmPassVisibility(true)}
                      className={classes.visibilityIcon}
                    />
                  ),
                }}
              />
              {!creatingUserLoader ? (
                <>
                  <Button style={{ color: "white" }} type="submit">
                    Create Account
                  </Button>

                  <Button
                    className={classes.cancelButton}
                    onClick={() => setModalOpen(false)}
                  >
                    Cancel
                  </Button>
                </>
              ) : (
                <>
                  <LinearProgress style={{ margin: "20px" }} />
                  <Button
                    className={classes.cancelButton}
                    onClick={() => setModalOpen(false)}
                  >
                    Cancel
                  </Button>
                </>
              )}
            </form>
          </div>
        </div>
      </div>
    </>
  );
};

export default Signup;
