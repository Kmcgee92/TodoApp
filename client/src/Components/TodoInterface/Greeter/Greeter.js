import React from "react";
//mui components
import Toolbar from "@material-ui/core/Toolbar";

const Greeter = ({ activeUser, modalOpen, setModalOpen, classes }) => {
  return (
    <>
      <Toolbar />
      {!Object.keys(activeUser).length && !modalOpen ? (
        <div data-test="greeting">
          <span className={classes.signupContent}>
            need to create an account?&nbsp;&nbsp;
            <span
              data-test="signupModalOpen"
              className={classes.signupToggleParent}
              onClick={() => setModalOpen(true)}
            >
              <span className={classes.signupToggle}>Signup</span>
            </span>
          </span>
        </div>
      ) : null}
    </>
  );
};

export default Greeter;
