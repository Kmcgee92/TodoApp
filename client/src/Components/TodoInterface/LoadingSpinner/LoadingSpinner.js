import React, {useState, useEffect} from "react";
import clsx from "clsx";
import CircularProgress from "@material-ui/core/CircularProgress";
import Fab from "@material-ui/core/Fab";
import CheckIcon from "@material-ui/icons/Check";
import CachedIcon from "@material-ui/icons/Cached";

export const LoadingSpinnerUX = ({classes, loading, setLoading, success, setSuccess}) => {

  const timer = React.useRef();

  useEffect(() => {
    setSuccess(false);
    setLoading(true);
    timer.current = window.setTimeout(() => {
      setSuccess(true);
      setLoading(false);
    }, 2000);
    return () => {
      clearTimeout(timer.current);
    };
  }, [setLoading, setSuccess]);

    const buttonClassname = clsx({
      [classes.buttonSuccess]: success,
      [classes.dataLoading]: !success,
    });
    
  return (
    <>
      <div className={classes.spinner}>
        <Fab aria-label="save" className={buttonClassname}>
          {success ? <CheckIcon /> : <CachedIcon />}
        </Fab>
        {loading && (
          <CircularProgress size={68} className={classes.fabProgress} />
        )}
      </div>
      <div className={classes.spinner}></div>
    </>
  );

}
const LoadingSpinner = ({ classes }) => {
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState(false);

  let props = {
    loading,
    setLoading,
    success,
    setSuccess,
    classes
  }
  return (
    <LoadingSpinnerUX {...props}/>
  );
};

export default LoadingSpinner;
