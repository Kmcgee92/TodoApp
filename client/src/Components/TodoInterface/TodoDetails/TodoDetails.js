import React, { useState, useEffect } from "react";
//redux
import { useDispatch, useSelector } from "react-redux";
import { updateTodos } from "../../../redux/actions/userTodoActions";
// apollo
import { useMutation } from "@apollo/react-hooks";
import { UPDATE_ITEM } from "../../../graphql/mutations/updateItem";

// mui core
import Typography from "@material-ui/core/Typography";
import Divider from "@material-ui/core/Divider";
import InputBase from "@material-ui/core/InputBase";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";

const TodoDetails = ({ classes, saving, setSaving }) => {
  const dispatch = useDispatch();
  const userId = useSelector((state) => state.auth.activeUser.id);
  const todoList = useSelector((state) => state.todos);
  const activeTodo = useSelector((state) => state.active);
  const [activeData, setActiveData] = useState({});
  const [complete, setComplete] = useState(activeData.completed || false);
  const [title, setTitle] = useState(activeData.title || "ADD TITLE");
  const [content, setContent] = useState(activeData.content || "");

  // eslint-disable-next-line
  let [saveData, { _data, _loading, _error }] = useMutation(UPDATE_ITEM);

  useEffect(() => {
    if (activeTodo) {
      const [filtered] = todoList.filter((object) => {
        return Number(object.id) === activeTodo;
      });
      setActiveData(filtered);
    }
  }, [todoList, activeTodo]);

  useEffect(() => {
    setComplete(activeData.completed);
    setTitle(activeData.title);
    setContent(activeData.content);
  }, [activeData]);

  const handleSave = async (e) => {
    setSaving(true);
    let completeFlag = complete;
    const target = e.target.innerHTML;
    if (target === "Incomplete" || target === "Complete") {
      setComplete(!complete);
      completeFlag = !complete;
    }
    await saveData({
      variables: {
        itemId: activeData.id,
        title: title || "ADD TITLE",
        content: content,
        completed: completeFlag !== "undefined" ? completeFlag : complete,
      },
    });
    const newTodo = {
      completed: completeFlag !== "undefined" ? completeFlag : complete,
      title: title || "ADD TITLE",
      content: content,
      id: activeData.id,
      userId: userId,
      __typename: "Item",
    };
    dispatch(updateTodos(newTodo));
  };

  const handleSetTitle = (e) => {
    setSaving(false);
    if (e.target.value.length < 80) {
      setTitle(e.target.value);
    } else return null;
  };
  const handleSetContent = (e) => {
    setSaving(false);
    setContent(e.target.value);
  };
  return (
    <div className={classes.detailContent}>
      <header className={classes.itemStatus}>
        {complete ? (
          <Button onClick={(e) => handleSave(e)} className={classes.complete}>
            Complete
          </Button>
        ) : (
          <Button onClick={(e) => handleSave(e)} className={classes.inComplete}>
            Incomplete
          </Button>
        )}
      </header>
      <Typography gutterBottom variant="h6">
        <InputBase
          name="title"
          autoComplete="off"
          autoFocus={true}
          className={classes.titleEditable}
          onChange={(e) => handleSetTitle(e)}
          onBlur={(e) => handleSave(e)}
          value={title || ""}
          inputProps={{ "aria-label": "naked" }}
        />
      </Typography>
      <Divider style={{ backgroundColor: "grey", marginBottom: "20px" }} />
      <Typography gutterBottom variant="h6">
        <TextField
          multiline={true}
          rows={30}
          InputProps={{
            className: classes.contentEditable,
            disableUnderline: true,
          }}
          autoComplete="off"
          name="content"
          fullWidth={true}
          onChange={(e) => handleSetContent(e)}
          onBlur={(e) => handleSave(e)}
          value={content || ""}
          inputProps={{ "aria-label": "naked" }}
        />
      </Typography>
    </div>
  );
};;

export default TodoDetails;

