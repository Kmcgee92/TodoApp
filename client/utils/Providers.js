import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
// provide redux and apollo
import { Provider } from "react-redux";
import { MockedProvider } from "@apollo/client/testing";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

export default function ComponentWithProviders(
  ComponentToBeTested,
  props,
  initialReduxMocks,
  initialApolloMocks
) {
  let mocks = initialApolloMocks || [];
  let store = mockStore(
    
    initialReduxMocks || { auth: { activeUser: {}, error: false } }
  
  );
  return (
    <MockedProvider mocks={mocks}>
      <Provider store={store}>
        <ComponentToBeTested {...props} />
      </Provider>
    </MockedProvider>
  );
}
