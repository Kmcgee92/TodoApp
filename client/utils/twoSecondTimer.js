export default function twoSecondTmer() {
  return new Promise((resolve) => {
    setTimeout(resolve, 2000);
  });
}
