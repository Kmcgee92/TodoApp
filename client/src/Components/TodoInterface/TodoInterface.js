import React, { useState, useEffect } from "react";
// redux
import { useSelector } from "react-redux";

// core components
import Header from './Header/Header'
import Signup from "../Auth/Signup";
import LoadingSpinner from "./LoadingSpinner/LoadingSpinner";
import TodoDetails from "./TodoDetails/TodoDetails";
import NoTasks from "./TodoDetails/NoTasks"
import TodoList from "./TodoList/TodoList";
import Greeter from "./Greeter/Greeter";
import SaveStatus from "./SaveStatus/SaveStatus";
//mui components
import Drawer from "@material-ui/core/Drawer";
import Toolbar from "@material-ui/core/Toolbar";

export function TodoInterface({ classes }) {
  const activeUser = useSelector((state) => state.auth.activeUser);
  const todoList = useSelector((state) => state.todos);
  const activeTodo = useSelector((state) => state.active);
  const [dataLoading, setDataLoading] = useState(false);
  const [modalOpen, setModalOpen] = useState(false);
  const [saving, setSaving] = useState(true);

  return (
    <div data-test="Root" className={classes.root}>
      <Header
        classes={classes}
        setDataLoading={setDataLoading}
        setModalOpen={setModalOpen}
      />
      <Drawer
        className={classes.drawer}
        variant="permanent"
        classes={{
          paper: classes.drawerPaper,
        }}
      >
        <Toolbar />
        <TodoList
          activeUser={activeUser}
          dataLoading={dataLoading}
          todoList={todoList}
          classes={classes}
        />
      </Drawer>
      <main className={classes.wrapper}>
        <Greeter
          activeUser={activeUser}
          modalOpen={modalOpen}
          setModalOpen={setModalOpen}
          classes={classes}
        />
        <SaveStatus
          activeUser={activeUser}
          activeTodo={activeTodo}
          classes={classes}
          saving={saving}
        />
        {(modalOpen || false) && (
          <Signup classes={classes} setModalOpen={setModalOpen} />
        )}
        {activeTodo && (
          <TodoDetails
            classes={classes}
            saving={saving}
            setSaving={setSaving}
          />
        )}
        {activeUser.items && !todoList.length && !dataLoading && (
          <NoTasks classes={classes} />
        )}
        {dataLoading && Object.keys(activeUser).length !== 0 && (
          <LoadingSpinner classes={classes} />
        )}
      </main>
    </div>
  );
}


export default TodoInterface;