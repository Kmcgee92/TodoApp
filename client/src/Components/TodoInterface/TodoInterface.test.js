import React from "react";
import { mount } from "enzyme";
// component to test
import TodoInterface from "./TodoInterface";
// child components
import Header from "./Header/Header";
import Signup from "../Auth/Signup";
import LoadingSpinner from "./LoadingSpinner/LoadingSpinner";
import TodoDetails from "./TodoDetails/TodoDetails";
import NoTasks from "./TodoDetails/NoTasks";
import TodoList from "./TodoList/TodoList";
import Greeter from "./Greeter/Greeter";
//mui child components
import Drawer from "@material-ui/core/Drawer";
import Toolbar from "@material-ui/core/Toolbar";
// utils
import { findByTestAtrr } from "../../../utils/index";
import Providers from "../../../utils/Providers";


describe("<TodoInterface/>", () => {
  let wrapWithProviders;
  let component;
  let props = {
    classes: {},
  };
  beforeEach(() => {
    wrapWithProviders = mount(Providers(TodoInterface, props));
    component = wrapWithProviders.find(TodoInterface);
  });

  describe("Should Render Component on Inital Mount", () => {
    beforeEach(() => {
      wrapWithProviders = mount(Providers(TodoInterface, props));
      component = wrapWithProviders.find(TodoInterface);
    });
    it('Should Match Snapshot', () => {
      expect(component).toMatchSnapshot()
    });
    it("Should Render Root Element on Mount", () => {
      const root = findByTestAtrr(component, "Root");
      expect(root).toHaveLength(1);
    });
    it("Should Render Header", () => {
      const header = component.find(Header);
      expect(header).toHaveLength(1);
    });
    it("Should Render Greeter", () => {
      const greeter = component.find(Greeter);
      expect(greeter).toHaveLength(1);
    });
    it("Should Render TodoList", () => {
      const todoList = component.find(TodoList);
      expect(todoList).toHaveLength(1);
    });
    it("Should Render Drawer", () => {
      const drawer = component.find(Drawer);
      expect(drawer).toHaveLength(1);
    });
    it("Should Render Three Toolbars on Full Mount in Greeter, Header, and the Interface", () => {
      const toolbars = component.find(Toolbar);
      expect(toolbars).toHaveLength(3);
    });
    it("Should not Render Signup, TodoDetails, NoTasks, or LoadingSpinner components on Mount", () => {
      let components = [];
      let array = [Signup, TodoDetails, NoTasks, LoadingSpinner];
      for (let i = 0; i < array.length; i++) {
        components.push(component.find(array[i]));
      }
      components.forEach((x) => {
        expect(x.length).toBe(0);
      });
    });
  });

  describe("Component Should Render with Props", () => {
    it("Should recieve any props passed in", () => {
      expect(component.props().classes).toEqual({});
    });
  });

  describe("Signup/TodoDetails/NoTasks Render on State Changes", ()=> {
    
    describe("Signup Component render", () => {
      it("Signup Button Function is successfully called on onClick event", () => {
        let mockChildProps = {
          activeUser: {}, 
          modalOpen: false, 
          setModalOpen: jest.fn(), 
          classes: {}
        }
        const childComponentGreeter = mount(<Greeter {...mockChildProps}/>)
        findByTestAtrr(childComponentGreeter, "signupModalOpen").simulate("click")
        expect(mockChildProps.setModalOpen).toBeCalledTimes(1)
      });

      it("Should Mount Signup Component Properly on State Change", () => {
        let signupSpan = findByTestAtrr(component, "signupModalOpen");
        signupSpan.simulate("click")
        //! because TodoInterface is wrapped with providers, ONLY root will update
        expect(wrapWithProviders.find(Signup)).toHaveLength(1)
      });
    })

    describe("TodoDetails Component Render", () => {
      it("Should Mount TodoDetails Component Properly on State Change", () => {
        let mockUser = {
          activeUser: {},
          error: false,
        };
        let mockTodos = [
          {
            id: 1,
            title: "mock",
            content: "mock",
            completed: false,
            __typename: "Item",
          },
        ];
        let mockReduxState = {
          active: 1,
          auth: mockUser,
          todos: mockTodos,
        };
        wrapWithProviders = mount(
          Providers(TodoInterface, props, mockReduxState)
        );
        component = wrapWithProviders.find(TodoInterface);
        expect(component.find(TodoDetails)).toHaveLength(1);
      });
    });

    describe("NoTasks Component Render", () => {
      it("Should Mount NoTasks Component Properly on State Change", () => {
        let mockReduxState = {
          auth: {
            activeUser: {
              items: "mockTRUE",
            },
            error: false,
          },
          todos: []
        };
  
        wrapWithProviders = mount(
          Providers(TodoInterface, props, mockReduxState)
          );
        component = wrapWithProviders.find(TodoInterface);
        expect(component.find(NoTasks)).toHaveLength(1);
      });
    })
    describe("LOADING SPINNER COMPONENT NEEDS A TEST FROM PARENT", ()=> {
      it("NEEDS TO HAVE A TEST FOR THE SET TIMEOUT ON dataLoading STATE", () => {
        expect(1).toBe(1)
      })
    })
  })
})
