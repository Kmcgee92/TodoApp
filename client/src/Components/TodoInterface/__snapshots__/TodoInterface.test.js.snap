// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoInterface/> Should Render Component on Inital Mount Should Match Snapshot 1`] = `ReactWrapper {}`;
